sys/kern/vm_object.c

```diff
@@ -21,21 +21,32 @@ vm_object_t *vm_object_alloc(vm_pgr_type_t type) {
   vm_object_t *obj = pool_alloc(P_VMOBJ, M_ZERO);
   TAILQ_INIT(&obj->list);
   RB_INIT(&obj->tree);
+  mtx_init(&obj->mtx, 0);
   obj->pager = &pagers[type];
+  obj->ref_counter = 1;
   return obj;
 }
 
 void vm_object_free(vm_object_t *obj) {
-  while (!TAILQ_EMPTY(&obj->list)) {
-    vm_page_t *pg = TAILQ_FIRST(&obj->list);
-    TAILQ_REMOVE(&obj->list, pg, obj.list);
-    vm_page_free(pg);
+  if (!refcnt_release(&obj->ref_counter)) {
+    return;
+  }
+
+  assert(obj->ref_counter == 0);
+
+  WITH_MTX_LOCK (&obj->mtx) {
+    while (!TAILQ_EMPTY(&obj->list)) {
+      vm_page_t *pg = TAILQ_FIRST(&obj->list);
+      TAILQ_REMOVE(&obj->list, pg, obj.list);
+      vm_page_free(pg);
+    }
   }
   pool_free(P_VMOBJ, obj);
 }
```

Comment:
Are we sure that nothing will change the value of `ref_counter` between `refcnt_release` and `WITH_MTX_LOCK`?

Reply:
We aren't. I will change it